File: flutter/lib/core/config/server_endpoint.dart
import 'dart:convert';

const _defaultRestUrl = 'https://api.devnu.ir/';
const _pinPrefix = 'chatnu-ca=';

class ChatNuServerEndpoint {
  const ChatNuServerEndpoint({required this.restUri, this.tlsCaPin});

  factory ChatNuServerEndpoint.production() =>
      ChatNuServerEndpoint.parse(_defaultRestUrl);

  factory ChatNuServerEndpoint.parse(String rawValue) {
    var raw = rawValue.trim();
    if (raw.isEmpty) {
      throw const FormatException('Server address is required.');
    }
    if (!raw.contains('://')) raw = 'https://$raw';
    final parsed = Uri.tryParse(raw);
    if (parsed == null || !parsed.hasAuthority || parsed.host.isEmpty) {
      throw const FormatException('Enter a valid HTTP or HTTPS server origin.');
    }
    if (parsed.scheme != 'https' && parsed.scheme != 'http') {
      throw const FormatException('Only HTTP and HTTPS servers are supported.');
    }
    if (parsed.userInfo.isNotEmpty || parsed.hasQuery) {
      throw const FormatException(
        'Server URLs cannot contain credentials or query parameters.',
      );
    }
    if (parsed.path.isNotEmpty && parsed.path != '/') {
      throw const FormatException(
        'Use the server origin only, for example https://chat.example.com.',
      );
    }

    String? pin;
    if (parsed.fragment.isNotEmpty) {
      if (!parsed.fragment.startsWith(_pinPrefix)) {
        throw const FormatException('Unsupported server enrollment fragment.');
      }
      pin = parsed.fragment.substring(_pinPrefix.length);
      if (!_validPin(pin)) {
        throw const FormatException('Invalid ChatNU emergency CA pin.');
      }
      if (parsed.scheme != 'https') {
        throw const FormatException(
          'Emergency pinned enrollment requires HTTPS.',
        );
      }
    }

    return ChatNuServerEndpoint(
      restUri: Uri(
        scheme: parsed.scheme,
        host: parsed.host,
        port: parsed.hasPort ? parsed.port : null,
        path: '/',
      ),
      tlsCaPin: pin,
    );
  }

  final Uri restUri;
  final String? tlsCaPin;

  String get restBaseUrl => restUri.toString();

  Uri get websocketUri => restUri.replace(
    scheme: restUri.scheme == 'https' ? 'wss' : 'ws',
    path: '/realtime',
  );

  String get identityNamespace {
    final port = restUri.hasPort
        ? restUri.port
        : restUri.scheme == 'https'
        ? 443
        : 80;
    return '${restUri.scheme}://${restUri.host}:$port';
  }

  String get hostLabel {
    final defaultPort = restUri.scheme == 'https' ? 443 : 80;
    final host = restUri.hasPort && restUri.port != defaultPort
        ? '${restUri.host}:${restUri.port}'
        : restUri.host;
    return tlsCaPin == null ? host : '$host · pinned';
  }

  String get enrollmentValue =>
      tlsCaPin == null ? restBaseUrl : '$restBaseUrl#$_pinPrefix$tlsCaPin';

  bool get usesEmergencyTls => tlsCaPin != null;

  static bool _validPin(String value) {
    if (!value.startsWith('sha256/')) return false;
    try {
      return base64.decode(value.substring('sha256/'.length)).length == 32;
    } on FormatException {
      return false;
    }
  }
}

File: app/src/main/java/com/example/remote/ServerEndpoint.kt
package com.example.remote

import android.content.Context
import com.example.BuildConfig
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

/**
 * Runtime-selected ChatNU instance.
 *
 * The endpoint is intentionally not secret, so it is stored in ordinary SharedPreferences.
 * Session credentials remain inside TokenStore and are protected by Android Keystore.
 */
object ServerEndpoint {
    private const val PREFS_NAME = "chatnu_server"
    private const val KEY_API_URL = "api_url"

    @Volatile
    private var selectedApiUrl: String = BuildConfig.CHATNU_API_URL

    fun initialize(context: Context) {
        val stored = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getString(KEY_API_URL, null)
        selectedApiUrl = normalize(stored ?: BuildConfig.CHATNU_API_URL)
            .getOrElse { BuildConfig.CHATNU_API_URL }
    }

    fun apiUrl(): String = selectedApiUrl

    fun hostLabel(): String = selectedApiUrl.toHttpUrlOrNull()?.let { url ->
        if (url.port == url.defaultPort()) url.host else "${url.host}:${url.port}"
    } ?: selectedApiUrl

    /** Namespace cryptographic account aliases so the same username on two servers is distinct. */
    fun identityNamespace(): String = selectedApiUrl.toHttpUrlOrNull()?.let { url ->
        "${url.scheme}://${url.host}:${url.port}"
    } ?: selectedApiUrl

    fun configure(context: Context, rawValue: String): Result<String> {
        return normalize(rawValue).map { normalized ->
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                .edit()
                .putString(KEY_API_URL, normalized)
                .apply()
            selectedApiUrl = normalized
            normalized
        }
    }

    fun reset(context: Context) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .remove(KEY_API_URL)
            .apply()
        selectedApiUrl = BuildConfig.CHATNU_API_URL
    }

    private fun normalize(rawValue: String): Result<String> = runCatching {
        var raw = rawValue.trim()
        require(raw.isNotBlank()) { "Server address is required." }
        if (!raw.contains("://")) raw = "https://$raw"

        val url = raw.toHttpUrlOrNull() ?: error("Enter a valid http(s) server address.")
        require(url.scheme == "https" || url.scheme == "http") { "Only HTTP or HTTPS servers are supported." }
        require(url.username.isEmpty() && url.password.isEmpty()) { "Server URLs cannot contain credentials." }
        require(url.query == null && url.fragment == null) { "Server URLs cannot contain a query or fragment." }
        require(url.encodedPath == "/" || url.encodedPath.isBlank()) {
            "Use the server origin only, for example https://chat.example.com"
        }

        url.newBuilder()
            .encodedPath("/")
            .query(null)
            .fragment(null)
            .build()
            .toString()
    }
}

private fun okhttp3.HttpUrl.defaultPort(): Int = when (scheme) {
    "https" -> 443
    else -> 80
}
